pkg/arvo/app/soto.hoon

```diff
@@ -0,0 +1,159 @@
+::  Soto: A Dojo relay for Urbit's Landscape interface
+::  Relays sole-effects to subscribers and forwards sole-action pokes
+/-  sole
+/+  *server, *sole
+/=  index
+  /^  octs
+  /;  as-octs:mimes:html
+  /:  /===/app/soto/index
+  /|  /html/
+      /~  ~
+  ==
+/=  tile-js
+  /^  octs
+  /;  as-octs:mimes:html
+  /:  /===/app/soto/js/tile
+  /|  /js/
+      /~  ~
+  ==
+/=  script
+  /^  octs
+  /;  as-octs:mimes:html
+  /:  /===/app/soto/js/index
+  /|  /js/
+      /~  ~
+  ==
+/=  style
+  /^  octs
+  /;  as-octs:mimes:html
+  /:  /===/app/soto/css/index
+  /|  /css/
+      /~  ~
+  ==
+/=  soto-png
+  /^  (map knot @)
+  /:  /===/app/soto/img  /_  /png/
+::
+=,  format
+|%
++$  state
+  $%  [%0 bon=bone]                                     :: store bone from Dojo peer
+  ==
+::
++$  move  [bone card]
+::
++$  poke
+  $%  [%launch-action [@tas path @t]]
+      [%sole-action sole-action:sole]
+  ==
+::
++$  card
+  $%  [%poke wire dock poke]
+      [%peer wire dock path]
+      [%http-response =http-event:http]
+      [%connect wire binding:eyre term]
+      [%diff diff]
+  ==
+::
++$  diff
+  $%  [%json json]
+      [%sole-effect sole-effect:sole]
+  ==
+::
+--
+::
+|_  [bol=bowl:gall sta=state]
+::
+++  this  .
+::
+++  prep
+  |=  old=(unit state)
+  ^-  (quip move _this)
+  =/  launcha=poke
+    [%launch-action [%soto /sototile '/~soto/js/tile.js']]
+  ?~  old
+    :_  this
+    :~ 
+        [ost.bol %connect / [~ /'~soto'] %soto]
+        [ost.bol %poke /soto [our.bol %launch] launcha]
+    ==
+  :-  [ost.bol %poke /soto [our.bol %launch] launcha]~
+  this(sta u.old)
+::
+++  peer-sototile
+  |=  wir=wire
+  ^-  (quip move _this)
+  :_  this
+  [ost.bol %diff %json *json]~
+::
+:: Peering Dojo when peered by front-end, initiating the session
+::
+++  peer-primary
+  |=  wir=wire
+  ^-  (quip move _this)
+  :-  [ost.bol %peer / [our.bol %dojo] /sole]~                                 :: poke/peer need same wire
+  %=  this
+    bon.sta  ost.bol
+  ==
+::
+++  poke-sole-action
+  |=  act=sole-action
+  ^-  (quip move _this)
+  :_  this
+  [bon.sta %poke / [our.bol %dojo] [%sole-action act]]~                        :: poke/peer need same wire
+::
```

Comment:
I recommend adding a poke-json that manually converts to poke-sole-action so as to improve speed.
Also, please assert that the poke is coming from (team:title our.bol src.bol).

Same comments for peer-primary.

Reply:
:/

does it at least work to keep the logic itself in the mark file, and pull that in as a "library", so the concerns don't get too conflated?

Reply:
No, the logic should be put into a library, which is then included by the mark. Akin to how chat does it

Reply:
> should

Marks are borked and we're using them in a borked way. This is on our radar and we'll give them some love at some point... but not this year!

Reply:
So like

```
++  poke-json
  |=  jon=json
  ^-  (quip move _this)
  :_  this
  [%sole-action jon]
```

? I can try and find examples in the codebase of this. Same with the `team:title` construction.

Reply:
More like

```hoon
++  poke-json
  |=  =json
  ^-  (quip move _this)
  (poke-action (json-to-action =json))
```